File: java/src/main/java/challenges/aoc2024/day2/function/ReactorSafetyPredicate.java
package challenges.aoc2024.day2.function;

import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import challenges.aoc2024.day2.model.ReactorReport;
import challenges.aoc2024.day2.model.type.GradientDirection;

public class ReactorSafetyPredicate implements Predicate<ReactorReport> {
    
    @Override
    public boolean test(ReactorReport reactorReport) {

        final GradientDirection direction = determineGradientDirection(reactorReport.getLevels());

        if (direction == null) {
            return false;
        }

        return IntStream.range(0, reactorReport.getLevels().size() - 1)
            .filter(i -> isSafeIncreaseOrDecrease(direction, reactorReport.getLevels().get(i), reactorReport.getLevels().get(i + 1)))
            .toArray().length + 1 == reactorReport.getLevels().size();
    }

    private GradientDirection determineGradientDirection(List<Integer> levels) {
        if (levels.stream().distinct().collect(Collectors.toList()).size() != levels.size()) {
            return null;
        }

        final List<GradientDirection> changes = IntStream.range(0, levels.size() - 1)
            .mapToObj(i -> levels.get(i) > levels.get(i + 1) ? GradientDirection.DECREASE : GradientDirection.INCREASE)
            .toList();

        if (changes.stream().distinct().collect(Collectors.toList()).size() > 1) {
            return null;
        }

        return changes.get(0);
    }

    private boolean isSafeIncreaseOrDecrease(GradientDirection direction, Integer startLevel, Integer nextLevel) {
        return (direction == GradientDirection.INCREASE && startLevel < nextLevel && startLevel + 3 >= nextLevel)
            || (direction == GradientDirection.DECREASE && startLevel > nextLevel && startLevel - 3 <= nextLevel);
    }
}

File: scala/src/main/scala/challenges/aoc2024/day2/function/ReactorSafetyPredicate.scala
package challenges.aoc2024.day2.function

import challenges.aoc2024.day2.model.ReactorReport
import challenges.aoc2024.day2.model.`type`.GradientDirection

class ReactorSafetyPredicate(isProblemDampenerActive: Boolean) extends (ReactorReport => Boolean) {

  override def apply(reactorReport: ReactorReport): Boolean = {

    val initiallySafe: Boolean = isSafe(reactorReport.levels)

    if !isProblemDampenerActive || initiallySafe then
      initiallySafe
    else
      reactorReport.levels.indices
        .map(indexToIgnore => reactorReport.levels.indices
          .filter(_.!=(indexToIgnore))
          .map(reactorReport.levels)
          .toList)
        .filter(isSafe)
        .toList.nonEmpty
  }

  private def isSafe(levels: List[Int]): Boolean = {

    try
      val direction = determineGradientDirection(levels).get
      levels
        .sliding(2)
        .collect { case Seq(a, b) => (a, b) }
        .forall((a, b) => isSafeIncreaseOrDecrease(direction, a, b))
    catch
      case e: NoSuchElementException => false
  }

  private def determineGradientDirection(levels: List[Int]): Option[GradientDirection] = {

    if levels.distinct.size != levels.size then
      None
    else
      val changes: List[GradientDirection] =
        levels.sliding(2).collect {
          case List(a, b) =>
            if a > b then GradientDirection.Decrease
            else GradientDirection.Increase
        }.toList

      changes.distinct match {
        case dir :: Nil => Some(dir)
        case _ => None
      }
  }

  private def isSafeIncreaseOrDecrease(direction: GradientDirection, startLevel: Int, nextLevel: Int): Boolean = {
    (direction == GradientDirection.Increase && startLevel < nextLevel && startLevel + 3 >= nextLevel)
    || (direction == GradientDirection.Decrease && startLevel > nextLevel && startLevel - 3 <= nextLevel)
  }
}
